File: s3/dsa/practice/stackLL.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <limits.h>

struct Node {
    struct Node* next;
    int x;
};

typedef struct Node Node;

struct LL {
    Node* head;
    int size;
};

typedef struct LL LL;

Node* getNode();
LL* getLL();
void push(LL* ll, int x);
void pop(LL* ll);
void display(LL* ll);

int main() {
    LL* ll = getLL();
    push(ll, 1); 
    display(ll);
    push(ll, 2);
    display(ll);
    push(ll, 5);
    display(ll);
    push(ll, 6);
    display(ll);
    pop(ll);
    display(ll);
    pop(ll);
    display(ll);
    push(ll, 7);
    display(ll);
}

Node* getNode() {
    Node* node = (Node*)malloc(sizeof(Node));
    node->next = NULL;
    return node;
}

LL* getLL() {
    LL* ll = (LL*)malloc(sizeof(LL));
    ll->head = NULL;
    ll->size = 0;
    return ll;
}

void push(LL* ll, int x) {
    Node* node = getNode();
    node->x = x;
    if(ll->head == NULL) {
        ll->head = node;
    }
    else {
        Node* tmp = ll->head;
        while(tmp->next != NULL) {
            tmp = tmp->next; 
        }
        tmp->next = node;
    }
    return;
}

void pop(LL *ll) {
    if(ll->head == NULL) {
        printf("Underflow; cannot pop\n");
    }
    else if(ll->head->next == NULL) {
        Node* tmp = ll->head;
        ll->head = NULL;
        free(tmp);
    }
    else {
        Node* tmp = ll->head;
        Node* prev = NULL;
        while(tmp->next != NULL) {
            prev = tmp;
            tmp = tmp->next;
        }
        free(tmp);
        prev->next = NULL;
    }
    return;
}

void display(LL* ll) {
    Node* tmp = ll->head;
    while(tmp != NULL) {
        printf("%d  ", tmp->x);
        tmp = tmp->next;
    }
    printf("\n");
}

File: s3/java/practice/stackLL.java
class Node {
    int x;
    Node next;

    Node() { this.next = null; }

    Node(int x) {
        this.x = x;
        this.next = null;
    }
}

class Stack {
    Node head;

    Stack() {
        head = new Node(Integer.MIN_VALUE);
        head.next = null;
    }

    void push(int x) {
        Node newNode = new Node(x);
        Node rover = head;
        while(rover.next != null) {
            rover = rover.next;    
        }
        rover.next = newNode;
        return;
    }

    void pop() {
        Node rover = head;
        Node parent = null;  
        while(rover.next != null) {
            parent = rover;
            rover = rover.next;
        }
        if(rover == head) {
            System.out.println("Stack underflow");
        }
        else {
            rover = null;
            parent.next = null;
        }
        return;
    }

    void display() {
        Node rover = head;
        while(rover.next != null) {
            rover = rover.next;
            System.out.print(rover.x);
        }
        System.out.println();
        return;
    }

    int top() {
        Node rover = head;
        while(rover.next != null) {
            rover = rover.next;
        }
        if(rover == head) {
            System.out.println("Stack underflow");
            return Integer.MIN_VALUE;
        }
        else {
            return rover.x;
        }
    }
}

class stackLL {
    public static void main(String args[]) {
        Stack stack = new Stack();
        stack.push(1);
        stack.push(2);
        stack.push(3);
        stack.push(4);
        stack.display();
        stack.pop();
        stack.display();
        System.out.println(stack.top());
        stack.pop();
        stack.pop();
        stack.pop();
        stack.top();
        stack.display();
        return;
    }
}
